src/passes/Flatten.cpp

```diff
@@ -192,6 +193,40 @@ struct Flatten
         loop->finalize();
         replaceCurrent(rep);
 
+      } else if (auto* tryy = curr->dynCast<Try>()) {
+        // remove a try value
+        Expression* rep = tryy;
+        auto* originalBody = tryy->body;
+        std::vector<Expression*> originalCatchBodies(tryy->catchBodies.begin(),
+                                                     tryy->catchBodies.end());
+        auto type = tryy->type;
+        Expression* prelude = nullptr;
+        if (type.isConcrete()) {
+          Index temp = builder.addVar(getFunction(), type);
+          if (tryy->body->type.isConcrete()) {
+            tryy->body = builder.makeLocalSet(temp, tryy->body);
+          }
+          for (Index i = 0; i < tryy->catchBodies.size(); i++) {
+            tryy->catchBodies[i] =
+              builder.makeLocalSet(temp, tryy->catchBodies[i]);
+          }
+          // the whole try is now a prelude
+          prelude = tryy;
+          // and we leave just a get of the value
+          rep = builder.makeLocalGet(temp, type);
+        }
+        tryy->body = getPreludesWithExpression(originalBody, tryy->body);
+        for (Index i = 0; i < tryy->catchBodies.size(); i++) {
+          tryy->catchBodies[i] = getPreludesWithExpression(
+            originalCatchBodies[i], tryy->catchBodies[i]);
+        }
+        tryy->finalize();
+        if (prelude) {
+          ReFinalizeNode().visit(prelude);
```

Comment:
if `prelude` is not null, isn't it equal to `tryy`? If so then we finalized it two lines above?

Reply:
Hmm, I think you're right. I kind of copy-pasted `if`'s code... Should I fix `if` too?

Reply:
By the way, are these any different?
```cpp
expr->finalize();
```
vs.
```cpp
ReFinalize().visit(expr);
```

Reply:
Ah, let's fix if too then, looks like it's also redundant there. Probably a copy-paste error there from back in history...

Yes, those two expressions are identical in practice, but the first uses static type info to call the right finalize function while the second is dynamic.

Reply:
Removed the second (redundant) finalization from both `try` and `if`.

Reply:
Hmm, turns out `if`'s finalization was necessary, because `if` has `condition` so `prelude` can be different from `iff`. Restored `if`'s finalization.